datafusion/core/tests/parquet_filter_pushdown.rs

```diff
@@ -0,0 +1,471 @@
+    expected_rows: usize,
+}
+
+impl TestCase {
+    fn new() -> Self {
+        Self {
+            // default to a filter that passes everything
+            filter: lit(true),
+            pushdown_expected: PushdownExpected::None,
+            expected_rows: 0,
+        }
+    }
+
+    /// Set the filter expression to use
+    fn with_filter(mut self, filter: Expr) -> Self {
+        self.filter = filter;
+        self
+    }
+
+    /// Set the expected predicate pushdown
+    fn with_pushdown_expected(mut self, pushdown_expected: PushdownExpected) -> Self {
+        self.pushdown_expected = pushdown_expected;
+        self
+    }
+
+    /// Set the number of expected rows (to ensure the predicates have
+    /// a good range of selectivity
+    fn with_expected_rows(mut self, expected_rows: usize) -> Self {
+        self.expected_rows = expected_rows;
+        self
+    }
+
+    async fn run(&self) {
+        // Also try and reorder the filters
+        // aka if the filter is `A AND B`
+        // this code will also try  `B AND A`
+        let filters = split_conjunction(&self.filter);
+
+        for perm in filters.iter().permutations(filters.len()) {
+            let perm: Vec<Expr> = perm.iter().map(|e| (**e).clone()).collect();
+            let filter = conjunction(perm).expect("had at least one conjunction");
+            self.run_with_filter(&filter).await
+        }
+    }
+
+    /// Scan the parquet file with the filters with various pushdown options
+    async fn run_with_filter(&self, filter: &Expr) {
+        let no_pushdown = self
+            .read_with_options(
+                &TESTFILE,
+                ParquetScanOptions {
+                    pushdown_filters: false,
+                    reorder_filters: false,
+                    enable_page_index: false,
+                },
+                // always expect no pushdown
+                PushdownExpected::None,
+                filter,
+            )
+            .await;
+
+        let only_pushdown = self
+            .read_with_options(
+                &TESTFILE,
+                ParquetScanOptions {
+                    pushdown_filters: true,
+                    reorder_filters: false,
+                    enable_page_index: false,
+                },
+                self.pushdown_expected,
+                filter,
+            )
+            .await;
+
+        assert_eq!(no_pushdown, only_pushdown);
+
+        let pushdown_and_reordering = self
+            .read_with_options(
+                &TESTFILE,
+                ParquetScanOptions {
+                    pushdown_filters: true,
+                    reorder_filters: true,
+                    enable_page_index: false,
+                },
+                self.pushdown_expected,
+                filter,
+            )
+            .await;
+
+        assert_eq!(no_pushdown, pushdown_and_reordering);
+
+        // page index filtering is not correct:
```

Comment:
I think this can be enabled now?

Reply:
I agree -- in https://github.com/apache/arrow-datafusion/pull/4062